// snoostorm/out/main.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Poll_1 = __importDefault(require("./util/Poll"));
exports.Poll = Poll_1.default;
exports.DefaultOptions = {
    pollTime: 2000,
    limit: 50,
    subreddit: "all",
};
var CommentStream = /** @class */ (function (_super) {
    __extends(CommentStream, _super);
    function CommentStream(client, options) {
        if (options === void 0) { options = exports.DefaultOptions; }
        var _this = _super.call(this, {
            frequency: options.pollTime || 2000,
            get: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                return [2 /*return*/, client.getNewComments(options.subreddit, options)];
            }); }); },
            identifier: "id",
        }) || this;
        return _this;
    }
    return CommentStream;
}(Poll_1.default));
exports.CommentStream = CommentStream;
var SubmissionStream = /** @class */ (function (_super) {
    __extends(SubmissionStream, _super);
    function SubmissionStream(client, options) {
        if (options === void 0) { options = exports.DefaultOptions; }
        var _this = _super.call(this, {
            frequency: options.pollTime || 2000,
            get: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                return [2 /*return*/, client.getNew(options.subreddit, options)];
            }); }); },
            identifier: "id",
        }) || this;
        return _this;
    }
    return SubmissionStream;
}(Poll_1.default));
exports.SubmissionStream = SubmissionStream;
var InboxStream = /** @class */ (function (_super) {
    __extends(InboxStream, _super);
    function InboxStream(client, options) {
        if (options === void 0) { options = {
            filter: "inbox",
            pollTime: 2000,
            limit: 5,
        }; }
        var _this = _super.call(this, {
            frequency: options.pollTime,
            get: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                return [2 /*return*/, client.getInbox(options)];
            }); }); },
            identifier: "id",
        }) || this;
        return _this;
    }
    return InboxStream;
}(Poll_1.default));
exports.InboxStream = InboxStream;
var ModMailStream = /** @class */ (function (_super) {
    __extends(ModMailStream, _super);
    function ModMailStream(client, options) {
        if (options === void 0) { options = exports.DefaultOptions; }
        var _this = _super.call(this, {
            frequency: options.pollTime || 2000,
            get: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                return [2 /*return*/, client.getModmail(options)];
            }); }); },
            identifier: "id",
        }) || this;
        return _this;
    }
    return ModMailStream;
}(Poll_1.default));
exports.ModMailStream = ModMailStream;
var ModQueueStream = /** @class */ (function (_super) {
    __extends(ModQueueStream, _super);
    function ModQueueStream(client, options) {
        var _this = _super.call(this, {
            frequency: options.pollTime || 2000,
            get: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                return [2 /*return*/, client.getSubreddit(options.subreddit).getModqueue(options)];
            }); }); },
            identifier: "id",
        }) || this;
        return _this;
    }
    return ModQueueStream;
}(Poll_1.default));
exports.ModQueueStream = ModQueueStream;
